Index: json_tokener.c
===================================================================
--- json_tokener.c	(revision 39)
+++ json_tokener.c	(working copy)
@@ -98,7 +98,7 @@
   tok->err = json_tokener_success;
 }
 
-struct json_object* json_tokener_parse(char *str)
+struct json_object* json_tokener_parse(const char *str)
 {
   struct json_tokener* tok;
   struct json_object* obj;
@@ -179,7 +179,7 @@
 
 
 struct json_object* json_tokener_parse_ex(struct json_tokener *tok,
-					  char *str, int len)
+					  const char *str, int len)
 {
   struct json_object *obj = NULL;
   char c = '\1';
Index: json_tokener.h
===================================================================
--- json_tokener.h	(revision 39)
+++ json_tokener.h	(working copy)
@@ -87,9 +87,9 @@
 extern struct json_tokener* json_tokener_new(void);
 extern void json_tokener_free(struct json_tokener *tok);
 extern void json_tokener_reset(struct json_tokener *tok);
-extern struct json_object* json_tokener_parse(char *str);
+extern struct json_object* json_tokener_parse(const char *str);
 extern struct json_object* json_tokener_parse_ex(struct json_tokener *tok,
-						 char *str, int len);
+						 const char *str, int len);
 
 #ifdef __cplusplus
 }
Index: Makefile.am
===================================================================
--- Makefile.am	(revision 39)
+++ Makefile.am	(working copy)
@@ -1,4 +1,4 @@
-AM_CFLAGS = -Wall -std=gnu99 -D_GNU_SOURCE -D_REENTRANT
+AM_CFLAGS = -Wall -Wwrite-strings -Werror -std=gnu99 -D_GNU_SOURCE -D_REENTRANT
 
 EXTRA_DIST = doc
 
